### PR TITLE
Delete and redeploy object upon error 'field is immutable'

### PR DESCRIPTION
**Pull request by priyawadhwa**
As discussed in #891, when running skaffold dev certain immutable Kubernetes
objects (like Jobs) can't be redeployed. A 'field is immutable' error is
returned when this happens.

To fix this issue, we can check the error from kubectl apply for 'field
is immutable'. If we find it, we can delete the object and try to deploy
it again.

Adds an integration test for skaffold dev (#441)

### CODE REVIEW

File: pkg/skaffold/deploy/kubectl/cli.go [hunk 1]
```diff
@@ -56,8 +59,25 @@ func (c *CLI) Apply(ctx context.Context, out io.Writer, manifests ManifestList)
...
-		return nil, errors.Wrap(err, "kubectl apply")
+	buf := bytes.NewBuffer([]byte{})
+	writer := bufio.NewWriter(buf)
+	if err := c.Run(ctx, manifests.Reader(), writer, "apply", c.Flags.Apply, "-f", "-"); err != nil {
+		if !strings.Contains(buf.String(), "field is immutable") {
+			return nil, err
+		}
+		// If the output contains the string 'field is immutable', we want to delete the object and recreate it
+		// See Issue #891 for more information
+		if err := c.Detete(ctx, out, manifests); err != nil {
```
Comment (r2d4) [inline pkg/skaffold/deploy/kubectl/cli.go]: this will need a rebase on balint's PR

**Comment by priyawadhwa** [comment]
@balopat for sure, I added one!

**Comment by dgageot** [comment]
@priyawadhwa should we switch to applying object one by one so that if only one can't be recreated, we don't force delete the others?

File: pkg/skaffold/deploy/kubectl/cli.go [hunk 2]
```diff
@@ -58,28 +58,29 @@ func (c *CLI) Apply(ctx context.Context, out io.Writer, manifests ManifestList)
...
-			return nil, errors.Wrap(err, "kubectl apply after deletion")
-		}
-	} else {
-		// Write output to out
-		if _, err := out.Write(buf.Bytes()); err != nil {
-			return nil, errors.Wrap(err, "writing to out")
+	for _, mfst := range manifests {
+		buf := bytes.NewBuffer([]byte{})
+		writer := bufio.NewWriter(buf)
+		ml := ManifestList{mfst}
```
Comment (r2d4) [inline pkg/skaffold/deploy/kubectl/cli.go]: no need to convert this back to a manifest list, mfst is just a []byte at this point

**Comment by dgageot** [comment]
Is applying one by one slower? How much?

**Comment by priyawadhwa** [comment]
Timings with this PR:
```
--- PASS: TestRun (180.37s)
    --- PASS: TestRun/getting-started_example (26.92s)
    --- PASS: TestRun/annotated_getting-started_example (8.94s)
    --- PASS: TestRun/getting-started_envTagger (14.37s)
    --- PASS: TestRun/gcb_builder_example (31.13s)
    --- PASS: TestRun/deploy_kustomize (4.46s)
    --- PASS: TestRun/bazel_example (47.35s)
    --- PASS: TestRun/kaniko_example (35.48s)
    --- PASS: TestRun/helm_example (11.73s)
--- PASS: TestDev (13.37s)
    --- PASS: TestDev/delete_and_redeploy_job (13.37s)
--- PASS: TestFix (14.83s)
PASS
```

A previous run:
```
--- PASS: TestRun (156.64s)
    --- PASS: TestRun/getting-started_example (18.55s)
    --- PASS: TestRun/annotated_getting-started_example (7.16s)
    --- PASS: TestRun/getting-started_envTagger (7.25s)
    --- PASS: TestRun/gcb_builder_example (29.19s)
    --- PASS: TestRun/deploy_kustomize (3.27s)
    --- PASS: TestRun/bazel_example (48.78s)
    --- PASS: TestRun/kaniko_example (33.81s)
    --- PASS: TestRun/helm_example (8.62s)
--- PASS: TestFix (16.07s)
PASS
```

So it does take a few seconds more for all of the tests except for the bazel test. Since this change only applies to developing with certain objects (Jobs, CronJobs) maybe it would be better to delete all objects and recreate instead?

**Comment by priyawadhwa** [comment]
Update: @r2d4 suggested we use the `--force` flag with `kubectl apply` which will delete and recreate an object if patching fails:

```
  --force=false: Delete and re-create the specified resource, when PATCH encounters conflict and has retried for 5 times.
```

Support for using this flag with the immutable error was added in this [PR](https://github.com/kubernetes/kubernetes/commit/0af6faed09535ba2763c07e25ac9402cd1f7270e), which hasn't been released officially. As discussed offline, we'll wait for kubectl v1.12.0 to come out so that we can use this flag.

**Comment by balopat** [comment]
One more comment: we should be careful about version support for kubectl and make it clear in the docs what works. Also we could print a warning for lower than supported kubectl versions.

**Comment by priyawadhwa** [comment]
kubectl v1.12.0 is out, this should be RFAL :)

**Comment by priyawadhwa** [comment]
@dgageot so the flag exists in lower versions it just doesn't work as expected, so if a user has a lower version then the same bug from the issue will occur.

File: docs/concepts.adoc [hunk 3]
```diff
@@ -28,6 +28,8 @@ tools for deployment, for example `kubectl` or `helm`.
 Each deployment type has parameters that allow you to
 define how you want your app to be installed and updated.
 
+_Note: kubectl version 1.12.0 or greater is recommended for use with skaffold._
```
[the remark this answers is not in the document]
Comment (priyawadhwa) [reply docs/concepts.adoc]: for sure, would this be the right spot for that?

**Comment by albertkang** [comment]
Any ideas when this will get merged?

File: docs/concepts.adoc (hunk 3 above)
[the remark this answers is not in the document]
Comment (nkubala) [reply docs/concepts.adoc]: ah, sorry missed your reply on this. FWIW this is a fine place to do it :)

**Comment by haf-afa** [comment]
Shouldn't it be the same for `skaffold run`? Because for me that errors.

**Comment by haf** [comment]
Ping, @priyawadhwa  @dgageot , see the comment from last year: this is still broken

**Comment by priyawadhwa** [comment]
hey @haf -- what version of kubectl are you using? This feature only works with kubectl >1.12; if you're already on that, would you mind opening an issue so this bug can be tracked?

**Comment by balopat** [comment]
hi @haf - you can override the behavior for `skaffold run` with `--force` - does that work for you?

**Comment by haf** [comment]
I’m using latest of everything. 

I never do force on almost anything. Only interested in happy path for the team.

**Pull request merged**